### DISCUSSION TITLE
[Bug] Discounts are applied even when they shouldn't

### DISCUSSION THREAD

**Discussion by mewa** [swellstores/community#432] [Help]
I raised this issue back in August and I just had another instance of it happen.

It is possible for a customer to qualify for a discount, apply it  (regardless if it's a store-wide promotion or a coupon code), then remove qualifying conditions and the customer is STILL able to checkout with the discount applied.

Example: store-wide discount, when buying 2 items you get -10% off the second item.
<img width="413" alt="Zrzut ekranu 2023-11-7 o 00 55 42" src="https://github.com/swellstores/community/assets/7347194/3ea6a2d5-faeb-422d-b6b1-30bc5642f5a8">

However, when you remove the qualifying (2nd) item, **the discount is still applied to the cart AND it stays like this throughout the checkout**
<img width="437" alt="Zrzut ekranu 2023-11-7 o 00 56 05" src="https://github.com/swellstores/community/assets/7347194/304eb4a5-4541-4c5c-bfe9-b54971741002">

There are a few flavours of this bug. Another one is when combined with shipping.

In the example below, customer had order of 108 EUR, which qualified them for free shipping (for orders of 100 EUR and above). However, they also used a coupon code, which brought their order below 100 EUR. Swell checkout still charged them with the free shipping and the discounted amount, but back **in the dashboard it shows that it did so incorrectly (see the difference between order total and paid)**.
![checkout-discount](https://github.com/swellstores/community/assets/7347194/007689b5-fbeb-4833-a45e-faf888b0633c)

It seems the core of the issue is that discounts are calculated at the wrong time.

**Comment by SusannaSwell** [comment]
hello @mewa , could you please share your 'Buy X get Y' coupon settings? I tried to reproduce, but without success. 
as of the second issue, we're investigating

**Reply by mewa** [reply]
Sure @SusannaSwell.

I thought it might be due to multiple 'Buy X get Y' settings – because the calculations are off if there are multiple discounts in a promo – but even the simplest config like the one below triggers it.
![2023-11-08-135130_1359x831_scrot](https://github.com/swellstores/community/assets/7347194/9fef4aa7-1365-4d2a-bf03-8dd1ee3792c6)

Using these steps in `wasabi-hq-eu` store using such config reproduces the bug without fail:
1. Add 2 *different* products from the X category
2. Observe discount applied
3. Remove the more expensive product
4. Observe discount is still applied

I had to disable this test promo but you can reproduce the same behaviour in the store by following the same principles (so add at least 3 chopsticks, of which 2 have to be a different product) to the promo visible on the product page.

This doesn't happen when adding & removing quantities of the same product. It also doesn't happen if you remove the cheaper product from the cart.

Weirdly enough, the same config and steps used in a test store don't trigger this behaviour.

**Reply by SusannaSwell** [reply]
Thank you for the explanations, coupon settings look okay. I tried to reproduce by adding 2 different products by one quantity of each, then remove the more expensive product and it works on my end. 
But you state that you're adding 3 items: quantity 2 of one product and quantity 1  of another product. Then you delete 1 quantity, two items seem to stay in cart, in this case rules of coupon are still valid. as it counts quantity of items from the chosen category

**Reply by mewa** [reply]
Sorry, I may have introduced some confusion, so let me clarify – the rule you're seeing in the screenshot is for a test promo, that I simplified as much as possible to reproduce the behaviour. Since it's a live store that accepts orders I cannot leave it on like that for an extended period of time – but the same principle (just the numbers are different) applies for the promotion which is live on the website – i.e. "buy 3, get -30% off the 3rd item".

**Reply by SusannaSwell** [reply]
in this case, when you need to use this scenario "buy 3, get -30% off the 3rd item"
You need to set into Customers BUYS quantity **3**
and into customer GETS amount 30% and up to quantity **1**

**Comment by mewa** [comment]
Was there a recent release that fixed this? It's still pretty much an issue
